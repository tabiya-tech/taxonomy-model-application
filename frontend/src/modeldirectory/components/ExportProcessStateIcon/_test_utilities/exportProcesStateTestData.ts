import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllExportProcessStatePermutations = (): ModelInfoTypes.ExportProcessState[] => {
  // generate all permutations of ExportProcessState status, errored, exportErrors, exportWarnings
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ExportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((exportErrors) => {
        allBooleans.forEach((exportWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              exportErrors: exportErrors,
              exportWarnings: exportWarnings,
            },
            downloadUrl: "https://example.com",
            timestamp: new Date(),
          });
        });
      });
    });
  });
  return allPermutations;
};
