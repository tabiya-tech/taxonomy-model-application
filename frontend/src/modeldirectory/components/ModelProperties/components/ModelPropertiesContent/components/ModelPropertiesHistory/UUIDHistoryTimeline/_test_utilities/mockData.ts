import { ModelInfoTypes } from "src//modelInfo/modelInfoTypes";

export function getOneFakeUUIDHistoryDetail(i: number): ModelInfoTypes.ModelHistory {
  return {
    UUID: "e26aca13-aa00-4346-9e83-edaf59314378" + i,
    id: "5e12f537f079fbdcb4e61bf" + i,
    localeShortCode: "ZA",
    name: "Model" + i,
    version: "V1.1." + i,
  };
}

export function getFakeUUIDHistoryDetailsArray(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => getOneFakeUUIDHistoryDetail(i));
}

export function getFakeUUIDHistoryDetailsArrayWithUnresolvedUUIDS(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => {
    const details = getOneFakeUUIDHistoryDetail(i);
    details.id = null;
    details.name = null;
    details.version = null;
    details.localeShortCode = null;
    return details;
  });
}
