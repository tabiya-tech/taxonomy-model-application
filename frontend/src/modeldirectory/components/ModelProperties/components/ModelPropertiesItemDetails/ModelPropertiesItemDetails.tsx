import { Box, Typography } from "@mui/material";
import React from "react";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import CustomIconButton from "src/theme/IconButton/CustomIconButton";
import { ContentCopy } from "@mui/icons-material";

export interface ModelPropertiesItemDetailsProps {
  title: string;
  value: string;
  isCopyEnabled?: boolean;
  "data-testid"?: string;
}

const uniqueId = "5abb0d55-4073-4f5b-8432-ee8c96aa4df2";
export const DATA_TEST_ID = {
  ITEM_TITLE: `item-title-${uniqueId}`,
  ITEM_VALUE: `item-value-${uniqueId}`,
  ITEM_COPY_BUTTON: `item-copy-button-${uniqueId}`,
};

const ModelPropertiesItemDetails = (props: Readonly<ModelPropertiesItemDetailsProps>) => {
  const { enqueueSnackbar } = useSnackbar();
  const handleCopyToClipboard = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    navigator.clipboard.writeText(props.value).then(() => {
      enqueueSnackbar("Copied to clipboard", { variant: "success", preventDuplicate: true });
    });
  };

  return (
    <Box data-testid={props["data-testid"]} display="flex" flexDirection="column" sx={{ width: "100%" }}>
      <Typography variant="h6" data-testid={DATA_TEST_ID.ITEM_TITLE} tabIndex={0}>
        {props.title}
      </Typography>
      <Box display="flex" flexDirection="row" alignItems={"flex-start"} justifyContent={"space-between"}>
        <Typography
          variant="body1"
          display="flex"
          justifyContent="space-between"
          alignItems="center"
          data-testid={DATA_TEST_ID.ITEM_VALUE}
        >
          {props.value}
        </Typography>
        <Box>
          {props.isCopyEnabled && (
            <CustomIconButton
              onClick={handleCopyToClipboard}
              title={`Copy ${props.title} to clipboard`}
              data-testid={DATA_TEST_ID.ITEM_COPY_BUTTON}
              sx={{ padding: (theme) => theme.tabiyaSpacing.xs }}
            >
              <ContentCopy />
            </CustomIconButton>
          )}
        </Box>
      </Box>
    </Box>
  );
};

export default ModelPropertiesItemDetails;
