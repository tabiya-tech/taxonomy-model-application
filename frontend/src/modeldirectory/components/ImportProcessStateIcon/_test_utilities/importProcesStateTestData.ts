import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllImportProcessStatePermutations = (): ModelInfoTypes.ImportProcessState[] => {
  // generate all permutations of ImportProcessState status, errored, parsingErrors, parsingWarnings
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ImportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((parsingErrors) => {
        allBooleans.forEach((parsingWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              parsingErrors: parsingErrors,
              parsingWarnings: parsingWarnings,
            },
            createdAt: new Date(allPermutations.length),
            updatedAt: new Date(allPermutations.length),
          });
        });
      });
    });
  });
  return allPermutations;
};
