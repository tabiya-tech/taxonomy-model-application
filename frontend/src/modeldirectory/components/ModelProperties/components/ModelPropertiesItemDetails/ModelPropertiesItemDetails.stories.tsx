import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import ModelPropertiesItemDetails from "./ModelPropertiesItemDetails";
import ModelInfoAPISpecs from "api-specifications/modelInfo";
import { getRandomLorem } from "src/_test_utilities/specialCharacters";

const meta: Meta = {
  title: "ModelDirectory/ModelProperties/ModelPropertiesItemDetails",
  component: ModelPropertiesItemDetails,
  tags: ["autodocs"],
  args: {
    title: "Title",
    value:
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
  },
};

type Story = StoryObj<typeof ModelPropertiesItemDetails>;

export const Shown: Story = {};

export const CopyToClipboard: Story = {
  args: {
    isCopyEnabled: true,
  },
};
export const ShownWithLongText: Story = {
  args: {
    value: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
  },
};

export const ShownWithLongTextAndCopyToClipboard: Story = {
  args: {
    value: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
    isCopyEnabled: true,
  },
};

export const Multiple: Story = {
  render: renderMultiple,
};

export const MultipleAndCopyToClipboard: Story = {
  render: renderMultiple,
  args: {
    isCopyEnabled: true,
  },
};

function renderMultiple(args: any) {
  const itemCount = 5;
  return (
    <React.Fragment>
      {Array.from({ length: itemCount }).map((_, index) => (
        <ModelPropertiesItemDetails
          key={`item-${index}`}
          {...args}
          title={`Title ${index}`}
          value={getRandomLorem(Math.floor(Math.random() * 100 + 10))}
        />
      ))}
    </React.Fragment>
  );
}

export default meta;
