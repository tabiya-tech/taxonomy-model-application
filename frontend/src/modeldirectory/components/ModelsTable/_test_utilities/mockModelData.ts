import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { ImportProcessStateEnums } from "api-specifications/importProcessState/enums";

import { v4 as uuidv4 } from "uuid";

import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { getMockId } from "src/_test_utilities/mockMongoId";
import { getRandomLorem, getRandomString, getTestString } from "src/_test_utilities/specialCharacters";
import { faker } from "@faker-js/faker";
import { CELL_MAX_LENGTH } from "src/modeldirectory/components/ModelsTable/ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  const model = getArrayOfFakeModels(1)[0];
  model.id = getMockId(id ?? 1);
  return model;
}

export function getArrayOfFakeModels(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location
          .countryCode("alpha-3")
          .substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description:
        i % 2 === 0
          ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH)
          : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getArrayOfFakeModelsMaxLength(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export const fakeModel: ModelInfoTypes.ModelInfo = {
  UUID: "8d914eab-6f7d-4183-accc-a09f99887b39",
  UUIDHistory: ["8d914eab-6f7d-4183-accc-a09f99887b39"],
  description: "aw j zt   agrkasl dy ogtimpsauwumu l utrovthao syertm beawpxluhyudgzbbm",
  id: "000000000000000000000001",
  importProcessState: {
    id: "000000000000000000000001",
    result: {
      errored: true,
      parsingErrors: true,
      parsingWarnings: true,
    },
    status: ImportProcessStateEnums.Status.PENDING,
    createdAt: new Date("2023-10-18T17:35:10.571Z"),
    updatedAt: new Date("2023-10-18T17:35:12.571Z"),
  },
  exportProcessState: [
    {
      downloadUrl: "https://www.example.com/",
      id: "000000000000000000000001",
      result: {
        errored: true,
        exportErrors: true,
        exportWarnings: true,
      },
      status: ExportProcessStateAPISpecs.Enums.Status.PENDING,
      timestamp: new Date("2023-10-18T17:35:10.571Z"),
      createdAt: new Date("2023-10-18T17:35:10.571Z"),
      updatedAt: new Date("2023-10-18T17:35:12.571Z"),
    },
  ],
  locale: {
    UUID: "7c4ea00b-fbea-4a31-9c04-009c813f5f34",
    name: "Palestine",
    shortCode: "ETH",
  },
  name: "1/1 - fysmros tcgcnjbbrrev",
  path: "https://unwelcome-editorial.net/",
  releaseNotes: "Perferendis modi impedit necessitatibus a",
  released: true,
  tabiyaPath: "https://silver-vulture.biz",
  createdAt: new Date("2023-10-18T17:35:10.571Z"),
  updatedAt: new Date("2022-10-18T17:35:10.571Z"),
  version: "0.2.2",
};

export function getRandomImportStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getRandomExportStatus(id: number) {
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getOneFakeExportProcessState(i: number): ModelInfoTypes.ExportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomExportStatus(i),
    result: {
      errored: i % 2 === 0,
      exportErrors: i % 2 === 0,
      exportWarnings: i % 2 === 0,
    },
    downloadUrl: faker.internet.url(),
    timestamp: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}

export function getOneFakeImportProcessState(i: number): ModelInfoTypes.ImportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomImportStatus(i),
    result: {
      errored: i % 2 === 0,
      parsingErrors: i % 2 === 0,
      parsingWarnings: i % 2 === 0,
    },
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}
