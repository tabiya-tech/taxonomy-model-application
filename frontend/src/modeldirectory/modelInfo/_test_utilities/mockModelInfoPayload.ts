import { faker } from '@faker-js/faker';

import * as ModelInfo from "api-specifications/modelInfo";
import {v4 as uuidv4} from "uuid";

import {getMockId} from "src/_test_utilities/mockMongoId";
import {getRandomString, getTestString} from "src/_test_utilities/specialCharacters";

export function getMockModelInfoPayload(number: number): ModelInfo.Types.GET.Response.Payload {
  return Array.from({length: number}, (_, i) => {
    return {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: getRandomString(ModelInfo.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(ModelInfo.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(ModelInfo.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: getTestString(ModelInfo.Constants.DESCRIPTION_MAX_LENGTH),
      released: false,
      releaseNotes: getTestString(ModelInfo.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfo.Constants.VERSION_MAX_LENGTH),
      // @ts-ignore
      createdAt: new Date().toISOString(),
      // @ts-ignore
      updatedAt: new Date().toISOString(),
      path: "https://foo/bar",
      tabiyaPath: "https://foo/bar/baz"
    }
  });
}

export function getFakeModelInfoPayload(number: number): ModelInfo.Types.GET.Response.Payload {
  return Array.from({length: number}, (_, i) => {
    return {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: faker.lorem.text().substring(0, ModelInfo.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country(),
        shortCode: faker.lorem.text().substring(0,  ModelInfo.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: faker.lorem.text().substring(0, ModelInfo.Constants.DESCRIPTION_MAX_LENGTH),
      released: false,
      releaseNotes: faker.lorem.text().substring(0, ModelInfo.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.lorem.text().substring(0, ModelInfo.Constants.VERSION_MAX_LENGTH),
      // @ts-ignore
      createdAt: faker.date.anytime().toISOString(),
      // @ts-ignore
      updatedAt: faker.date.anytime().toISOString(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url()
    }
  });
}