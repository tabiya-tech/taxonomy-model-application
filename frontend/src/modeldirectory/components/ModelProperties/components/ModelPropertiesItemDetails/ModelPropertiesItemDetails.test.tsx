// mute the console
import "src/_test_utilities/consoleMock";

import ModelPropertiesItemDetails, { DATA_TEST_ID } from "./ModelPropertiesItemDetails";
import { fireEvent, screen, waitFor, within } from "@testing-library/react";
import { render } from "src/_test_utilities/test-utils";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";

// mock the snackbar
jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

describe("ItemDetails", () => {
  describe("render tests", () => {
    test("should render with provided title and value", () => {
      // GIVEN an item with a title and a value
      const givenItem = {
        title: "foo",
        value: "bar",
      };

      const givenDataTestId = "item-details";

      // WHEN the component is rendered
      render(<ModelPropertiesItemDetails {...givenItem} data-testid={givenDataTestId} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the component to be displayed
      const itemDetailsElement = screen.getByTestId(givenDataTestId);
      expect(itemDetailsElement).toBeInTheDocument();
      // AND the given item's title to be displayed
      const titleElement = within(itemDetailsElement).getByTestId(DATA_TEST_ID.ITEM_TITLE);
      expect(titleElement).toBeInTheDocument();
      expect(titleElement).toHaveTextContent(givenItem.title);
      // AND the given item's value to be displayed
      const valueElement = within(itemDetailsElement).getByTestId(DATA_TEST_ID.ITEM_VALUE);
      expect(valueElement).toBeInTheDocument();
      expect(valueElement).toHaveTextContent(givenItem.value);
      // AND the copy to clipboard button not to be displayed
      const copyButtonElement = within(valueElement).queryByTestId(DATA_TEST_ID.ITEM_COPY_BUTTON);
      expect(copyButtonElement).not.toBeInTheDocument();
      // AND the component to match the snapshot
      expect(itemDetailsElement).toMatchSnapshot();
    });

    test("should display a copy to clipboard button when the isCopyEnabled prop is true", () => {
      // GIVEN an item with a title and a value
      const givenItem = {
        title: "foo",
        value: "bar",
      };

      const givenDataTestId = "item-details";

      // WHEN the component is rendered with the isCopyEnabled prop set to true
      render(<ModelPropertiesItemDetails {...givenItem} isCopyEnabled={true} data-testid={givenDataTestId} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the component to be displayed
      const itemDetailsElement = screen.getByTestId(givenDataTestId);
      expect(itemDetailsElement).toBeInTheDocument();
      // AND the given item's title to be displayed
      const titleElement = within(itemDetailsElement).getByTestId(DATA_TEST_ID.ITEM_TITLE);
      expect(titleElement).toBeInTheDocument();
      expect(titleElement).toHaveTextContent(givenItem.title);
      // AND the given item's value to be displayed
      const valueElement = within(itemDetailsElement).getByTestId(DATA_TEST_ID.ITEM_VALUE);
      expect(valueElement).toBeInTheDocument();
      expect(valueElement).toHaveTextContent(givenItem.value);
      // AND the copy to clipboard button to be displayed
      const copyButtonElement = within(itemDetailsElement).getByTestId(DATA_TEST_ID.ITEM_COPY_BUTTON);
      expect(copyButtonElement).toBeInTheDocument();
      // AND the component to match the snapshot
      expect(itemDetailsElement).toMatchSnapshot();
    });
  });

  describe("action tests", () => {
    test("should copy the value to the clipboard when the copy to clipboard button is clicked", async () => {
      // Ensure navigator.clipboard is defined since jest/jsdom does natively implement it
      Object.defineProperty(navigator, "clipboard", {
        value: {
          writeText: jest.fn(),
        },
      });
      // Define writeText as a function if it's not already defined
      if (!navigator.clipboard.writeText) {
        navigator.clipboard.writeText = jest.fn();
      }
      const mockClipboardWriteText = jest.spyOn(navigator.clipboard, "writeText").mockResolvedValue(undefined);

      // GIVEN an item with a title and a value
      const givenItem = {
        title: "foo",
        value: "bar",
      };
      // AND the component is rendered with the isCopyEnabled prop set to true
      render(<ModelPropertiesItemDetails {...givenItem} isCopyEnabled={true} data-testid={"item-details"} />);

      // WHEN the copy to clipboard button is clicked
      const copyButtonElement = screen.getByTestId(DATA_TEST_ID.ITEM_COPY_BUTTON);
      fireEvent.click(copyButtonElement);

      // THEN expect the value to be copied to the clipboard
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(givenItem.value);
      await waitFor(() => expect(mockClipboardWriteText).toHaveBeenCalledWith(givenItem.value));
      await waitFor(() => {
        // AND expect a snackbar with a message to be shown
        expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith("Copied to clipboard", {
          variant: "success",
          preventDuplicate: true,
        });
      });
    });
  });
});
