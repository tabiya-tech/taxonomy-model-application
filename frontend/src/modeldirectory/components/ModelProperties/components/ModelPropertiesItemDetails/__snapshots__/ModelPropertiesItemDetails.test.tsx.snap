// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetails render tests should display a copy to clipboard button when the isCopyEnabled prop is true 1`] = `
<div
  class="MuiBox-root css-1ofqig9"
  data-testid="item-details"
>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-yyitj4-MuiTypography-root"
    data-testid="item-title-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    tabindex="0"
  >
    foo
  </h6>
  <div
    class="MuiBox-root css-16j8xl6"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1cy5wtl-MuiTypography-root"
      data-testid="item-value-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    >
      bar
    </p>
    <div
      class="MuiBox-root css-0"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1s9eyn2-MuiButtonBase-root-MuiIconButton-root"
        data-testid="item-copy-button-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
        tabindex="0"
        title="Copy foo to clipboard"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`ItemDetails render tests should render with provided title and value 1`] = `
<div
  class="MuiBox-root css-1ofqig9"
  data-testid="item-details"
>
  <h6
    class="MuiTypography-root MuiTypography-h6 css-yyitj4-MuiTypography-root"
    data-testid="item-title-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    tabindex="0"
  >
    foo
  </h6>
  <div
    class="MuiBox-root css-16j8xl6"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-1cy5wtl-MuiTypography-root"
      data-testid="item-value-5abb0d55-4073-4f5b-8432-ee8c96aa4df2"
    >
      bar
    </p>
    <div
      class="MuiBox-root css-0"
    />
  </div>
</div>
`;
